Build static hydra-node and hydra-tui executables

ch1bo opened the pull request:
:snowflake: Add nix expressions to compile hydra-node and hydra-tui as static executables in the Hydra CI (the nix thing)

:snowflake: Decreased size of our docker images by using static binaries and `alpine` as base  image.

![image](https://user-images.githubusercontent.com/2621189/195404092-f60d0a48-2396-4979-96d1-1240ee7a4862.png)

:snowflake: Added a Github CI job to build and upload static executables into an artifact

![image](https://user-images.githubusercontent.com/2621189/195450909-45fba503-0993-41ae-9811-161c177d34d5.png)
![image](https://user-images.githubusercontent.com/2621189/195450987-b631f53c-1da0-4164-be25-2e142895e2da.png)


:snowflake: Mentioned our use of static binaries in user manual & contribution guidelines.

ch1bo committed:
Add a hydra-node-static target to release.nix

This derivation uses haskell.nix's projectCross with musl64 pkgs.

ch1bo committed:
Add a static build for hydra-tui

This is using a patch from @angerman and @abailly

ch1bo committed:
Make libsodium override local to pkgs and patch terminfo only on musl64 package set

ch1bo committed:
Bump haskell.nix and iohk.nix

New haskell.nix does correctly propagate nonReinstallablePkgs to projectCross

New iohk.nix avoid deprecation warning on libsodium override

ch1bo committed:
Use not so up-to-date haskell.nix

Seems like one haskell package (moo) was tripping it (or it's cabal) as
it includes a module twice in moo.cabal

ch1bo committed:
Only reinstall terminfo when cross compiling

abailly commented:
If we have static binaries couldn't we use a scratch image for docker instead of alpine?

angerman commented:
If you need help with this, let me know!

angerman commented:
> If we have static binaries couldn't we use a scratch image for docker instead of alpine?

Would be worth a try, I'm not familiar enough with hydra and what kind of OS level support it expects.

Fundamentally a musl build only requires syscalls. However stuff like dns lookups might still try to read /etc/resolv.conf (fair warning! This will fail on systems that decided to have a different layout -- and you get failed dbs resolutions). This is primarily due to the fact that muslc does hardcode some paths. Hence if you OS deviates from the assumptions muslc makes, the static binaries libc component might not work properly.

All that said, you might need an alpine base image if you run into this kind of issues.

ch1bo committed:
Use quite recent haskell.nix and use forked 'moo'

Haskel.nix's Cabal version was tripped by the duplicate module so we
need to use the forked version.

ch1bo committed:
Replace ws -> websocat as it has been dropped

ch1bo committed:
Add hydraw-static expression to release.nix

ch1bo committed:
Try to use static binaries in Dockerfile

ch1bo committed:
Bump docker actions

ch1bo committed:
Configure dependabot for github actions

ch1bo committed:
Build static executables in CI workflow

ch1bo committed:
Unicode all the things

ch1bo committed:
Update more github actions

ch1bo committed:
Also compile and upload hydra-tools static executable

ch1bo committed:
Document that we do create static binaries

ch1bo committed:
Mention static binaries in release workflow

ch1bo committed:
Update another github action

ch1bo committed:
Rename executables artifact

ch1bo commented on CONTRIBUTING.md:
```diff
@@ -84,13 +84,14 @@ trusted-public-keys = cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDS
 - Run `haskell-language-server` for an IDE experience
 - Run `hoogle` for symbol & documentation lookup
+- Build the `hydra-node` using nix: `nix-build release.nix -A hydra-node`
+- Build a statically linked `hydra-node` using nix: `nix-build release.nix -A hydra-node`
```
Yes. It's missing a `-static`

ch1bo replied:
Fixed it now

ch1bo committed:
Fix contributing guidelines

ch1bo committed:
Use upstream of moo, which got fixed (but not on hackage still)
